File: Leetcode/linkedlist/147_InsertionSortList/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode dummy = new ListNode(Integer.MIN_VALUE);
        ListNode temp = dummy, curr = head, next = head;
        while(curr!=null){
            temp = dummy;
            next = curr.next;
            curr.next = null;
            while(temp.next!=null && curr.val>temp.next.val){
                temp = temp.next;
            }
            curr.next = temp.next;
            temp.next = curr;
            curr = next;
        }
        return dummy.next;
    }
}

File: Leetcode/linkedlist/147_InsertionSortList/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode("-inf")
        temp = dummy
        curr = head
        next = head
        while curr:
            temp = dummy
            next = curr.next
            curr.next = None
            while temp.next and curr.val > temp.next.val:
                temp = temp.next
            curr.next = temp.next
            temp.next = curr
            curr = next
        return dummy.next
